candidate-agreement/src/table.rs

```diff
@@ -0,0 +1,986 @@
+								first: (old_candidate, old_sig.clone()),
+								second: (candidate, signature.clone()),
+							})),
+							None,
+						);
+					}
+				} else {
+					existing.proposal = Some((digest.clone(), signature.clone()));
+				}
+			}
+			Entry::Vacant(vacant) => {
+				vacant.insert(ValidatorData {
+					proposal: Some((digest.clone(), signature.clone())),
+					known_statements: HashSet::new(),
+				});
+
+				// TODO: seed validity votes with issuer here?
+				self.candidate_votes.entry(digest.clone()).or_insert_with(move || CandidateData {
+					group_id: group,
+					candidate: candidate,
+					validity_votes: HashMap::new(),
+					availability_votes: HashMap::new(),
+					indicated_bad_by: Vec::new(),
+				});
+			}
+		}
+
+		self.validity_vote(
+			context,
+			from,
+			digest,
+			ValidityVote::Issued(signature),
+		)
+	}
+
+	fn validity_vote(
+		&mut self,
+		context: &C,
+		from: C::ValidatorId,
+		digest: C::Digest,
+		vote: ValidityVote<C::Signature>,
+	) -> (Option<<C as ResolveMisbehavior>::Misbehavior>, Option<Summary<C::Digest, C::GroupId>>) {
+		let votes = match self.candidate_votes.get_mut(&digest) {
+			None => return (None, None), // TODO: queue up but don't get DoS'ed
+			Some(votes) => votes,
+		};
+
+		// check that this validator actually can vote in this group.
+		if !context.is_member_of(&from, &votes.group_id) {
+			let (sig, valid) = match vote {
+				ValidityVote::Valid(s) => (s, true),
+				ValidityVote::Invalid(s) => (s, false),
+				ValidityVote::Issued(_) =>
+					panic!("implicit issuance vote only cast if the candidate entry already created successfully; qed"),
+			};
+
+			return (
+				Some(Misbehavior::UnauthorizedStatement(UnauthorizedStatement {
+					statement: SignedStatement {
+						signature: sig,
+						sender: from,
+						statement: if valid {
+							Statement::Valid(digest)
+						} else {
+							Statement::Invalid(digest)
+						}
+					}
+				})),
+				None,
+			);
+		}
+
+		// check for double votes.
+		match votes.validity_votes.entry(from.clone()) {
+			Entry::Occupied(occ) => {
+				if occ.get() != &vote {
+					let double_vote_proof = match (occ.get().clone(), vote) {
+						(ValidityVote::Issued(iss), ValidityVote::Valid(good)) |
+						(ValidityVote::Valid(good), ValidityVote::Issued(iss)) =>
+							ValidityDoubleVote::IssuedAndValidity((votes.candidate.clone(), iss), (digest, good)),
+						(ValidityVote::Issued(iss), ValidityVote::Invalid(bad)) |
+						(ValidityVote::Invalid(bad), ValidityVote::Issued(iss)) =>
+							ValidityDoubleVote::IssuedAndInvalidity((votes.candidate.clone(), iss), (digest, bad)),
+						(ValidityVote::Valid(good), ValidityVote::Invalid(bad)) |
+						(ValidityVote::Invalid(bad), ValidityVote::Valid(good)) =>
+							ValidityDoubleVote::ValidityAndInvalidity(digest, good, bad),
+						_ => {
+							// this would occur if two different but valid signatures
+							// on the same kind of vote occurred.
```

Comment:
we should consider making this a misbehaviour; it's intentionally confusing to sign multiple equivalent messages in different ways.

Reply:
I guess so. but it seems fairly benign and proper handling at the network layer will make floods ineffective in the first place.